[DFT] Implement external workspace according to draft spec

hjabird opened the pull request:
# Description

This PR implements the draft specification for external workspace at https://github.com/oneapi-src/oneAPI-spec/pull/509, which relates to oneMKL issue https://github.com/oneapi-src/oneMKL/issues/396.

# Checklist

## All Submissions

- [x] Do all unit tests pass locally? Attach a log. 
- [x] Have you formatted the code using clang-format?

## New interfaces

- [ ] Have you provided motivation for adding a new feature as part of RFC and
it was accepted? # (RFC) **Draft PR for oneAPI spec linked, but this is not yet merged**
- [ ] What version of oneAPI spec the interface is targeted? 
- [ ] Complete [New features](pull_request_template.md#new-features) checklist

## New features

- [x] Have you provided motivation for adding a new feature?
- [x] Have you added relevant tests?

Logs:
* [inteltest.txt](https://github.com/oneapi-src/oneMKL/files/13473985/inteltest.txt)
* [testamd.txt](https://github.com/oneapi-src/oneMKL/files/13473986/testamd.txt)
* [testnvidia.txt](https://github.com/oneapi-src/oneMKL/files/13473989/testnvidia.txt)

hjabird committed:
[DFT] External workspace implementation & tests

* Implementation for supporting external workspaces according to draft specification
  * Supporting MKL-GPU, rocFFT, cuFFT.
  * MKL-CPU needs minor changes, but does not support external workspaces
* Initial tests

hjabird committed:
Support resetting descriptors w/ test

FMarno commented on include/oneapi/mkl/dft/detail/external_workspace_helper.hpp:
```diff
@@ -0,0 +1,178 @@
+#include "oneapi/mkl/dft/detail/types_impl.hpp"
+#include "oneapi/mkl/dft/detail/commit_impl.hpp"
+
+namespace oneapi {
+namespace mkl {
+namespace dft {
+namespace detail {
```
```suggestion
namespace oneapi::mkl::dft::detail {
```
for me, this makes it clearer that there is only actually 1 workspace here so I don't need to worry about finding the brace where it ends.

hjabird replied:
Done in [bee7d30](https://github.com/oneapi-src/oneMKL/pull/411/commits/bee7d3068a28fa90a868f23baab198dffad3440b).

FMarno commented on include/oneapi/mkl/dft/detail/external_workspace_helper.hpp:
```diff
@@ -0,0 +1,178 @@
+public:
+    /** Constructor.
+     *  @param ext_workspace_rqd True if WORKSPACE_PLACEMENT is set to WORKSPACE_EXTERNAL.
+     *  @param backend_rqd_bytes_fn The function to call to obtain the backend's required workspace size.
```
```suggestion

```

hjabird replied:
Removed in [1da13ee](https://github.com/oneapi-src/oneMKL/pull/411/commits/1da13eef0fbb663fcadc1abbcc15ad930f977589).

Rbiessy commented on include/oneapi/mkl/dft/detail/commit_impl.hpp:
```diff
@@ -67,6 +77,20 @@ class commit_impl {
+        return external_workspace_helper_.get_rqd_workspace_bytes(*this);
+    };
+
+    // Set workspace by be overrided for any backend that actually uses the workspace.
```
Not sure what you meant here, there is a typo I think?

hjabird replied:
I've reworded this slightly in [1da13ee](https://github.com/oneapi-src/oneMKL/pull/411/commits/1da13eef0fbb663fcadc1abbcc15ad930f977589).

FMarno commented on include/oneapi/mkl/dft/detail/external_workspace_helper.hpp:
```diff
@@ -0,0 +1,178 @@
+    /** Reset. May be required for changing descriptor settings before recommitting.
+     *  @param ext_workspace_rqd True if WORKSPACE_PLACEMENT is set to WORKSPACE_EXTERNAL.
+    */
+    inline void reset(bool ext_workspace_rqd) {
```
was `reset` required when you had the `backend_rqd_bytes_fn` param?

hjabird replied:
Sorry, I don't understand the comment.

When a descriptor is recommitted, if there is an external workspace, user needs to:
* Re-query the external workspace size requirements.
* set_workspace (even if it had previously been set).

`reset()` resets the internal state of the workspace helper to make sure oneMKL throws the appropriate errors.

FMarno replied:
Sorry, I should have been clearer.
I was thinking users of the class could just do this action themselves, as in override their existing `external_workspace_helper` using the constructor and copy assignment operator.

hjabird replied:
Done in [1da13ee](https://github.com/oneapi-src/oneMKL/pull/411/commits/1da13eef0fbb663fcadc1abbcc15ad930f977589).

FMarno commented on include/oneapi/mkl/dft/detail/external_workspace_helper.hpp:
```diff
@@ -0,0 +1,178 @@
+    // Set workspace type, with optional workspaces.
+    ext_workspace_type m_workspace_type;
+
+    // Minimum size of workspace in bytes. -1 indicates not set.
+    std::int64_t m_workspace_bytes_rqd;
```
Why are we "caching" this?

hjabird replied:
Because:
* We need to provide this to the user, and also internally when set_workspace is called.
* We don't know the cost of asking the backend for the workspace space requirement. 
* We only need to obtain this value once per commit, so caching it is dead simple.

FMarno replied:
It may also be super cheap to query!
I was just think it might be unneeded complexity. Its very neatly contained so I'm not so bothered really.

FMarno commented on include/oneapi/mkl/dft/detail/external_workspace_helper.hpp:
```diff
@@ -0,0 +1,178 @@
+    inline void set_workspace_throw(commit_impl_t& committed_desc, scalar_t* usm_workspace) {
+        if (get_rqd_workspace_bytes(committed_desc) > 0 && usm_workspace == nullptr) {
+            throw mkl::invalid_argument("DFT", "set_workspace",
+                                        "Provided workspace is too small (nullptr)");
```
I think it would be better to clearly say that the function was expecting a non-null pointer.

hjabird replied:
The backend might not require an external workspace, in which case the user might legally provide a `nullptr`.

FMarno replied:
I'm just commenting on the explanation string. "too small" seems odd to me.
```suggestion
                                        "Backend expected a non-null workspace");
```

hjabird replied:
Done in [1da13ee](https://github.com/oneapi-src/oneMKL/pull/411/commits/1da13eef0fbb663fcadc1abbcc15ad930f977589).

FMarno commented on include/oneapi/mkl/dft/detail/external_workspace_helper.hpp:
```diff
@@ -0,0 +1,178 @@
+        if (get_rqd_workspace_bytes(committed_desc) > 0 && usm_workspace == nullptr) {
+            throw mkl::invalid_argument("DFT", "set_workspace",
+                                        "Provided workspace is too small (nullptr)");
+            return;
```
surely the return is unreachable?

hjabird replied:
Removed `return` in [1da13ee](https://github.com/oneapi-src/oneMKL/pull/411/commits/1da13eef0fbb663fcadc1abbcc15ad930f977589).

FMarno replied:
There are a number of other instances of this

hjabird replied:
I've removed more instances in [bee7d30](https://github.com/oneapi-src/oneMKL/pull/411/commits/bee7d3068a28fa90a868f23baab198dffad3440b).

Rbiessy commented on src/dft/backends/cufft/commit.cpp:
```diff
@@ -44,15 +44,20 @@ namespace detail {
     std::array<std::optional<cufftHandle>, 2> plans = { std::nullopt, std::nullopt };
     std::array<std::int64_t, 2> offsets;
+    // For using a buffer external workspace, the buffer object itself needs to be kept between compute calls such that accessors can be used.
+    std::optional<sycl::buffer<scalar_type, 1>> buffer_workspace;
```
Is this needed? I don't think it's being used since it is already part of `commit_impl` and `external_workspace_helper`

hjabird replied:
Its leftover from where my initial attempts didn't neatly fit into the existing code. Removed in [1da13ee](https://github.com/oneapi-src/oneMKL/pull/411/commits/1da13eef0fbb663fcadc1abbcc15ad930f977589).

Rbiessy commented on src/dft/backends/cufft/commit.cpp:
```diff
@@ -323,6 +351,76 @@ class cufft_commit final : public dft::detail::commit_impl<prec, dom> {
         return offsets;
     }
 
+    virtual void set_workspace(scalar_type* usmWorkspace) override {
```
I think we should stick to snake_case for function arguments, same for `bufferWorkspace`.

hjabird replied:
That would be consistent with the contribution guidelines for oneMKL. I've updated your suggestions here, and a few other places I missed in [1da13ee](https://github.com/oneapi-src/oneMKL/pull/411/commits/1da13eef0fbb663fcadc1abbcc15ad930f977589).

Rbiessy replied:
There are a few more occurrences of `bufferWorkspace` and `usmWorkspace` in other files.

hjabird replied:
Hopefully all the stragglers caught in [fd528ea](https://github.com/oneapi-src/oneMKL/pull/411/commits/fd528eae5034fec9ca50f349c74f8c7b3925f443).

FMarno commented on include/oneapi/mkl/dft/detail/external_workspace_helper.hpp:
```diff
@@ -0,0 +1,178 @@
+     *  @param committed_desc The backend's native descriptor.
+     *  @param buffer_workspace A buffer for the workspace
+    */
+    inline void set_workspace_throw(commit_impl_t& committed_desc,
```
Do you really need to mark the functions inline? This is a templated class so I wouldn't think ODR is an issue (could be wrong)

hjabird replied:
Removed `inline` in several places in [1da13ee](https://github.com/oneapi-src/oneMKL/pull/411/commits/1da13eef0fbb663fcadc1abbcc15ad930f977589).

Rbiessy commented on src/dft/backends/cufft/commit.cpp:
```diff
@@ -323,6 +351,76 @@ class cufft_commit final : public dft::detail::commit_impl<prec, dom> {
+    }
+
+    virtual std::int64_t get_workspace_external_bytes_impl() override {
+        std::int64_t size = plans[1] ? get_plan_workspace_size_bytes(*plans[1]) : 0;
+        return size + get_plan0_workspace_size_bytes();
```
I didn't think of the issue of multiple plans with the external workspace. I think we should return the maximum of the 2 workspace size. I'm concerned that this would artificially increase the workspace size for users that only use one direction.
This would add the requirement that we must wait between `compute` functions of a same descriptor that uses an external workspace which I think is acceptable.

hjabird replied:
I'd prefer not to add a user requirement about waits between compute calls. But I think we can avoid this:
* I think the buffer implementation already will already have implicit dependencies between compute forward / compute backwards, since they're getting access to the same underlying buffer. 
  * I kept it simple doing this. Anyone migrating from CUDA probably isn't using a buffer anyway.
* For USM we can always have a dependency on the last use of the allocation.
  * I don't *think* it can cause hidden deadlocks. Open to thoughts on that though.

Finally this might be good because in the implementation of this PR I introduced errors in both rocFFT and cuFFT when splitting up the buffers/USM allocs. 

Are you happy with the idea of implicit dependencies in the USM impl?

Rbiessy replied:
Yes that makes sense. As you say it should work for buffers and we can add the dependency ourselves for USM.
I think we could also add a comment somewhere or make a note in the spec that using external workspace serializes the computations of a same descriptor.

FMarno replied:
I think the synchronisation should be offloaded to the user. In much the same way that the user shouldn't try to launch two kernels that use the same USM memory without the use of events to synchronize, then it should be clear from our specification that the external workspace is in use while the `compute_forward` kernel is in flight.
Kernels launched from the same descriptor should be independent.
Trying to sort this ourselves goes against USM I would say.

hjabird replied:
I've implemented this with implicit USM dependencies in [9c2b2ff](https://github.com/oneapi-src/oneMKL/pull/411/commits/9c2b2ffc008faf9b5da5a137fa53e2e74f935785) and [19869bf](https://github.com/oneapi-src/oneMKL/pull/411/commits/19869bf06518db4983183a24ca7f68f907f1c834).

I agree that this doesn't fit with the typical usage of USM. However, anyone porting from closed-source *oneMKL* and expecting their code to work with *cuFFT* and *rocFFT* would expect their code to just work without additional changes to dependency tracking.

I'm also under the impression that CUDA streams are in-order (perhaps I'm wrong here - I've never used CUDA directly). Consequently, kernels launched from the same sycl queue would not run concurrently anyway. *cuFFT* launches multiple kernels per DFT, so the USM dependency would enforce that kernels from different DFTs could not be interleaved. In this context, the implicit dependency for the USM implementation does not seem so bad.

FMarno replied:
I'm having trouble finding the docs about closed-source oneMKL synchronizing on the use of the workspace, do you have a link?

CUDA streams are in-order, but we change the stream for each call to `compute_xxxward` to allow transforms to happen concurrently.

I think launching N kernels from the same descriptor is similar to launching 1 kernel from each of N descriptors that share the same external workspace. If you had limited memory then you could save a lot by sharing the same workspace between all the descriptors.

hjabird replied:
The reference for closed-source oneMKL is [here](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2023-0/user-allocated-workspaces.html). It states that:
> The descriptor assumes that the workspace will be solely used for FFT, so the user must be careful not to use the same workspace area for other purposes in parallel or before finishing compute_forward/compute_backward (especially for USM as explicit synchronization is required).

My interpretation of this is that for anything other than FFTs, explicit synchronisation is required. Therefore we don't have to worry about synchronisation for multiple DFTs with a single descriptor. Perhaps @raphael-egan can comment?

Regarding CUDA streams, my understanding of how streams map to CUDA was wrong. Thanks for clearing that up.

FMarno replied:
Thanks for the link!
To me, that reads only as a reminder that the descriptor cannot detect if the memory is being used elsewhere. I don't see a guarantee of synchronization between calls to compute_xxxward.

raphael-egan replied:
> The reference for closed-source oneMKL is [here](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2023-0/user-allocated-workspaces.html). It states that:
> 
> > The descriptor assumes that the workspace will be solely used for FFT, so the user must be careful not to use the same workspace area for other purposes in parallel or before finishing compute_forward/compute_backward (especially for USM as explicit synchronization is required).
> 
> My interpretation of this is that for anything other than FFTs, explicit synchronisation is required. Therefore we don't have to worry about synchronisation for multiple DFTs with a single descriptor. Perhaps @raphael-egan can comment?
> 
> Regarding CUDA streams, my understanding of how streams map to CUDA was wrong. Thanks for clearing that up.

Answer given by email and shared with Cheetah team.

hjabird replied:
It looks as if a number of backends don't make any gaurantees regarding calling multiple *compute_XXX* functions. For example, I believe that calling multiple *cuFFT* compute functions without interdependencies using the same descriptor could lead to problems in the current oneMKL implementation since the SYCL runtime will schedule those without explicit dependencies to run concurrently by changing streams - see [how cuFFT doesn't suppor this on Stackoverflow](https://stackoverflow.com/questions/39030949/cufftsetstream-causes-garbage-output-am-i-doing-something-wrong).

The problem of not scheduling concurrent `compute_*` calls therefore has a larger scope than just for explict workspace implementation, and I believe this should be addressed at a oneAPI spec level.

I'll be removing the interdepencies from the USM version of the code. For SYCL spec conformance, using accessors on workspace buffers is still the correct thing to do, so these will still implicitly result in implicit synchronisation.

hjabird replied:
Removed dependencies for USM compute calls in [9856b7d](https://github.com/oneapi-src/oneMKL/pull/411/commits/9856b7d5d320c5807e7a17cefad32b3a5074af9f).

Rbiessy replied:
That looks good to me so I'll resolve this big discussion, thanks.

FMarno commented on include/oneapi/mkl/dft/detail/external_workspace_helper.hpp:
```diff
@@ -0,0 +1,178 @@
+        }
+    }
+
+    inline void get_workspace_buffer_access_if_rqd(const char* function_name, sycl::handler& cgh) {
```
What does this do?

hjabird replied:
Now renamed `get_workspace_accessor_if_rqd`, and also returning `std::optional<sycl::accessor>`.

For backends that don't natively support buffers (eg. anything that isn't SYCL), the workspace buffer dependency of a compute function must be added the SYCL dependency graph when the call is submitted to the queue. This is achieved by creating an accessor from te workspace buffer. This function does that if its needed (ie. a buffer workspace is needed, and its size is greater than zero), and also now returns this workspace accessor.

FMarno commented on include/oneapi/mkl/dft/detail/external_workspace_helper.hpp:
```diff
@@ -0,0 +1,178 @@
+            }
+            throw mkl::invalid_argument(
+                "DFT", function_name,
+                "Buffer external workspace must be used with buffer compute calls");
```
```suggestion
            if (m_workspace_buffer) {
                if (m_workspace_buffer->size()) {
                    m_workspace_buffer->template get_access<sycl::access::mode::read_write>(cgh);
                }

            } else {
            throw mkl::invalid_argument(
                "DFT", function_name,
                "Buffer external workspace must be used with buffer compute calls");
                }
```

hjabird replied:
Done in [1da13ee](https://github.com/oneapi-src/oneMKL/pull/411/commits/1da13eef0fbb663fcadc1abbcc15ad930f977589).

FMarno commented on src/dft/backends/cufft/backward.cpp:
```diff
@@ -69,6 +69,7 @@ ONEMKL_EXPORT void compute_backward(descriptor_type &desc,
 
     queue.submit([&](sycl::handler &cgh) {
         auto inout_acc = inout.template get_access<sycl::access::mode::read_write>(cgh);
+        commit->get_workspace_buffer_access_if_rqd("compute_backward", cgh);
```
I presume this is here to show that the buffer is being used here so should be added to the dependency graph, but it makes me a little nervous that there is no variable with the scope of the whole function.

hjabird replied:
Yes, this is about adding the buffer dependency to the dependency graph. I also get nervous about the interaction of native APIs and buffers.

My understanding is that an accessor is safe to access within the scope of the enclosing command group scope. I think that this means that the dependency on `cgh` gives it the correct lifetime for the dependecy graph.

Rbiessy commented on tests/unit_tests/dft/source/workspace_external_tests.cpp:
```diff
@@ -0,0 +1,358 @@
+    sycl::free(deviceBwd, sycl_queue);
+    sycl::free(workspace, sycl_queue);
+    return sanityCheckPasses ? !::testing::Test::HasFailure() : ::testing::Test::HasFailure();
+    ;
```
```suggestion
```

hjabird replied:
Done in [1da13ee](https://github.com/oneapi-src/oneMKL/pull/411/commits/1da13eef0fbb663fcadc1abbcc15ad930f977589).

Rbiessy commented on tests/unit_tests/dft/source/workspace_external_tests.cpp:
```diff
@@ -0,0 +1,358 @@
+    }
+
+    return sanityCheckPasses ? !::testing::Test::HasFailure() : ::testing::Test::HasFailure();
+    ;
```
```suggestion
```

hjabird replied:
Done in [1da13ee](https://github.com/oneapi-src/oneMKL/pull/411/commits/1da13eef0fbb663fcadc1abbcc15ad930f977589).

FMarno commented on include/oneapi/mkl/dft/detail/external_workspace_helper.hpp:
```diff
@@ -0,0 +1,178 @@
+     *  @param committed_desc The backend's native descriptor.
+     *  @param usm_workspace A USM allocation for the workspace. Assumed to be sufficeintly large.
+    */
+    inline void set_workspace_throw(commit_impl_t& committed_desc, scalar_t* usm_workspace) {
```
I don't really like the name `set_workspace_throw` since it doesn't set the workspace and hopefully doesn't throw.
I would prefer something like `check_workspace_is_valid` or just `check_workspace_valid` or maybe `check_workspace_size`.

hjabird replied:
I agree its not a great name! It sets the workspace information within the external_workspace_helper, and should throw anything required by the specification.

I've renamed this `set_workspace_metadata` since this isn't just a check that its valid, and as you say, ideally doesn't throw. Let me know if you think this is good enough. In [bee7d30](https://github.com/oneapi-src/oneMKL/pull/411/commits/bee7d3068a28fa90a868f23baab198dffad3440b).

FMarno commented on src/dft/backends/cufft/commit.cpp:
```diff
@@ -315,6 +332,17 @@ class cufft_commit final : public dft::detail::commit_impl<prec, dom> {
     }
 
+    static void apply_external_workspace_setting(cufftHandle& handle,
+                                                 const config_value& workspace_setting) {
```
```suggestion
    static void apply_external_workspace_setting(cufftHandle handle,
                                                 config_value workspace_setting) {
```
cufftHandle is just an `unsigned int` so should be passed by value, and `config_value` is something like that too.

hjabird replied:
Done in [1da13ee](https://github.com/oneapi-src/oneMKL/pull/411/commits/1da13eef0fbb663fcadc1abbcc15ad930f977589).

FMarno commented on src/dft/backends/cufft/commit.cpp:
```diff
@@ -323,6 +351,76 @@ class cufft_commit final : public dft::detail::commit_impl<prec, dom> {
+                cufftSetWorkArea(plan, workspace_native);
+            });
+        });
+        this->get_queue().wait_and_throw();
```
```suggestion
        }).wait_and_throw();
```

hjabird replied:
Done in [1da13ee](https://github.com/oneapi-src/oneMKL/pull/411/commits/1da13eef0fbb663fcadc1abbcc15ad930f977589).

hjabird committed:
Suggested minor changes

FMarno commented on tests/unit_tests/dft/include/parseval_check.hpp:
```diff
@@ -0,0 +1,81 @@
+ * @param rescale_forward A value to multiply the in data by.
+*/
+template <typename TypeFwd, typename TypeBwd>
+bool parseval_check(std::size_t dft_len, const TypeFwd* in, TypeBwd* out,
```
```suggestion
bool parseval_check(std::size_t dft_len, const TypeFwd* in, const TypeBwd* out,
```

hjabird replied:
Done in [bee7d30](https://github.com/oneapi-src/oneMKL/pull/411/commits/bee7d3068a28fa90a868f23baab198dffad3440b).

FMarno commented on tests/unit_tests/dft/source/workspace_external_tests.cpp:
```diff
@@ -0,0 +1,357 @@
+    using backward_t = std::complex<scalar_t>;
+
+    sycl::queue sycl_queue(*dev);
+    if (prec == precision::DOUBLE && !sycl_queue.get_device().has(sycl::aspect::fp64)) {
```
```suggestion
    if (prec == precision::DOUBLE && !sycl_queue.get_device().has(sycl::aspect::fp64)) {
                    std::cout << "Device does not support double precision." << std::endl;
```

hjabird replied:
Done in [bee7d30](https://github.com/oneapi-src/oneMKL/pull/411/commits/bee7d3068a28fa90a868f23baab198dffad3440b).

FMarno commented on tests/unit_tests/dft/source/workspace_external_tests.cpp:
```diff
@@ -0,0 +1,357 @@
+    // Generate data
+    std::vector<forward_t> host_fwd(static_cast<std::size_t>(dft_size));
+    std::size_t bwd_size =
+        dom == domain::COMPLEX ? dft_size : dft_size / 2 + 1; // TODO: Check this!
```
It seems correct to me!

hjabird replied:
TODO removed in [bee7d30](https://github.com/oneapi-src/oneMKL/pull/411/commits/bee7d3068a28fa90a868f23baab198dffad3440b).

FMarno replied:
I don't see the change

hjabird replied:
Well the change was in my editor... unsaved. Oops. Done in [aa0967a](https://github.com/oneapi-src/oneMKL/pull/411/commits/aa0967a974e68d6701c408969b43132fd50be684).

FMarno commented on src/dft/backends/cufft/commit.cpp:
```diff
@@ -323,6 +350,77 @@ class cufft_commit final : public dft::detail::commit_impl<prec, dom> {
+
+    // Separate workspaces are needed for the forward and backward plans. Because the plans[1]'s workspace is offset
+    // by the size of plans[0]'s workspace, we need to be able to get plans[0]'s size individually.
+    std::int64_t get_plan0_workspace_size_bytes() {
+        if (!plans[0]) {
+            return 0;
+        }
+        return get_plan_workspace_size_bytes(*plans[0]);
+    }
```
```suggestion
    std::int64_t get_plan_workspace_size_bytes(std::size_t dir) {
        if (!plans[dir]) {
            return 0;
        }
        return get_plan_workspace_size_bytes(*plans[dir]);
    }
```
I don't think it's much to add a `0` to your above call and then this can be reused for `get_workspace_external_bytes_impl`

hjabird replied:
Now that a workspace is shared for forwards and backwards in the *cuFFT* and *rocFFT* backends, this function no longer exists.

FMarno commented on src/dft/backends/cufft/commit.cpp:
```diff
@@ -323,6 +350,77 @@ class cufft_commit final : public dft::detail::commit_impl<prec, dom> {
+
+    std::int64_t get_plan_workspace_size_bytes(cufftHandle handle) {
+        std::size_t size = 0;
+        cufftGetSize(*plans[0], &size);
```
```suggestion
        cufftGetSize(handle, &size);
```

hjabird replied:
Done in [bee7d30](https://github.com/oneapi-src/oneMKL/pull/411/commits/bee7d3068a28fa90a868f23baab198dffad3440b)

FMarno commented on src/dft/backends/cufft/commit.cpp:
```diff
@@ -323,6 +350,77 @@ class cufft_commit final : public dft::detail::commit_impl<prec, dom> {
+                    buffer_workspace.template get_access<sycl::access::mode::read_write>(cgh, range,
+                                                                                         offset);
+                cgh.host_task([=](sycl::interop_handle ih) {
+                    auto stream = ih.get_native_queue<sycl::backend::ext_oneapi_cuda>();
+                    auto result = cufftSetStream(plan, stream);
+                    if (result != CUFFT_SUCCESS) {
+                        throw oneapi::mkl::exception(
+                            "dft/backends/cufft", "set_workspace",
+                            "cufftSetStream returned " + std::to_string(result));
+                    }
```
Why are you setting the stream? the stream is changed with each call to compute forward

hjabird replied:
No longer calling setStream in [bee7d30](https://github.com/oneapi-src/oneMKL/pull/411/commits/bee7d3068a28fa90a868f23baab198dffad3440b).

t4c1 commented on include/oneapi/mkl/dft/detail/commit_impl.hpp:
```diff
@@ -67,6 +77,20 @@ class commit_impl {
+
+    // set_workspace should be overrided for any backend that enables external workspaces.
+    // If these are overrided, get_workspace_external_bytes_impl must also be overrided.
+    // For backends that do not support external workspaces, these functions do not need to be overrided.
```
this comment could be improved by explaining what these defaults do if not overridden.

hjabird replied:
I've expanded the comment in [d31868a](https://github.com/oneapi-src/oneMKL/pull/411/commits/d31868aefc19b5b7dfc15f661f62cd54b122a72c).

t4c1 commented on include/oneapi/mkl/dft/detail/external_workspace_helper.hpp:
```diff
@@ -0,0 +1,170 @@
+        if (m_ext_workspace_rqd) {
+            if (m_workspace_buffer) {
+                if (m_workspace_buffer->size()) {
+                    m_workspace_buffer->template get_access<sycl::access::mode::read_write>(cgh);
```
What is the point of creating accessor if it is not used anywhere?

hjabird replied:
The *cuFFT* and *rocFFT* backends take their workspace arguments as pointers. Consequently, implementing the buffer API with then poses a few issues. 

The issue that this aims to resolve is that of the SYCL runtime knowing that a workspace buffer is used when a DFT accesses it via a captured pointer (captured in `set_workspace`). Here, we create an accessor out of the buffer (to create the expected kernel dependencies) but then access the buffer via the captured pointer instead of via the accessor itself.

t4c1 commented on src/dft/backends/cufft/commit.cpp:
```diff
@@ -323,6 +350,77 @@ class cufft_commit final : public dft::detail::commit_impl<prec, dom> {
+        }
+        if (plans[1]) {
+            cufftSetWorkArea(
+                *plans[1], usm_workspace + get_plan0_workspace_size_bytes() / sizeof(scalar_type));
```
Do we need separate workspaces for both plans? Does the spec allow forward and backward compute of a descriptor running concurrently?

hjabird replied:
This is discussed in https://github.com/oneapi-src/oneMKL/pull/411#discussion_r1399082325. The implementation has been changed to use a single shared workspace, but the discussion is ongoing, so please add your thoughts there!

hjabird committed:
Share workspace between forward and backward DFTs for rocFFT and cuFFT impls

t4c1 commented on src/dft/backends/cufft/commit.cpp:
```diff
@@ -323,6 +350,77 @@ class cufft_commit final : public dft::detail::commit_impl<prec, dom> {
+                    }
+                    auto workspace_native = reinterpret_cast<scalar_type*>(
+                        ih.get_native_mem<sycl::backend::ext_oneapi_cuda>(workspace_acc));
+                    cufftSetWorkArea(plan, workspace_native);
```
I do not think this is OK. SYCL runtime only expects the memory to be used within the host_task call. This could lead to multiple subtle bugs. So accessor should be constructed within the same host_task call that launches the cufft kernel.

hjabird replied:
You probably correct from the standpoint of the spec (although the interop spec for CUDA or AMD are not yet published). My understanding is that this should work with Codeplay's oneAPI for Nvidia GPUs and oneAPI for AMD GPUs.

hjabird replied:
I was able to do this more tidily than expected in [fe3273f](https://github.com/oneapi-src/oneMKL/pull/411/commits/fe3273f53bb6b3a0449e94b65bba4d77877a62ee).

hjabird committed:
Fix previous commit for AMD, NVidia

hjabird committed:
Correct variable naming style #2

hjabird committed:
Rename to add_buffer_workspace_dependency_if_rqd

hjabird committed:
Fix previous commit

hjabird committed:
Improve comment on default behaviour

FMarno commented on src/dft/backends/cufft/commit.cpp:
```diff
@@ -323,6 +350,63 @@ class cufft_commit final : public dft::detail::commit_impl<prec, dom> {
+        }
+    }
+
+    void set_buffer_workspace(cufftHandle plan, sycl::buffer<scalar_type>& buffer_workspace) {
```
Why do we need this function if we are going to set the workspace with each compute call using buffers?

hjabird replied:
I give my reasoning in the comment starting on line 364 - to allow internal memory to be freed (if required), and to allow cuFFT to generate any errors (I think its preferable to get errors for setting the workspace at this point in comparison to later when compute calls are made). I'm not sure what errors are likely. I'd guess the most likely is that the buffer can't be accessed on the Nvidia device.

FMarno replied:
Sorry, yes you did write that.

hjabird commented:
I've been having trouble with access to AMD hardware since the rebase, so test here is test data for MKL-CPU, MKL-GPU and Nvidia for now:
[testintel.txt](https://github.com/oneapi-src/oneMKL/files/13582967/testintel.txt)
[testnvidia.txt](https://github.com/oneapi-src/oneMKL/files/13582968/testnvidia.txt)

hjabird committed:
Support portFFT; Update tests for where configurations are not supported

hjabird commented:
This PR has been rebased:
* DFT domain has had portFFT added as a backend
* Additional work was required to support portFFT in the tests

Test results for Nvidia (cuFFT backend), Intel MKL-CPU and MKL-GPU, and portFFT on Intel are attached:
[testintel_mkl.txt](https://github.com/oneapi-src/oneMKL/files/13679187/testintel_mkl.txt)
[testintel_portfft.txt](https://github.com/oneapi-src/oneMKL/files/13679188/testintel_portfft.txt)
[testnvidia.txt](https://github.com/oneapi-src/oneMKL/files/13679189/testnvidia.txt)

Unfortunately I cannot currently access AMD hardware to test that the rocFFT backend. However, it has not required any changes since the last test results were provided.

anantsrivastava30 commented on include/oneapi/mkl/dft/detail/commit_impl.hpp:
```diff
@@ -67,6 +77,23 @@ class commit_impl {
+    };
+
+    // set_workspace should be overrided for any backend that enables external workspaces.
+    // If these are overrided, get_workspace_external_bytes_impl must also be overrided.
```
overridden perhaps instead of overrided.

hjabird replied:
Yes, my English was incorrect there. Corrected in [2c4c096](https://github.com/oneapi-src/oneMKL/pull/411/commits/2c4c0962030d285710c0faf485be289711ff36bf).

hjabird replied:
@anantsrivastava30 if you're happy with this (and everything else), it looks like I'm good to merge this PR!

anantsrivastava30 reviewed (approved):
LGTM!

Rbiessy commented on tests/unit_tests/main_test.cpp:
```diff
@@ -102,9 +102,9 @@ int main(int argc, char** argv) {
-                    if (dev.is_gpu() && plat.get_info<sycl::info::platform::name>().find(
-                                            "OpenCL") != std::string::npos)
-                        continue;
+                    // if (dev.is_gpu() && plat.get_info<sycl::info::platform::name>().find(
+                    //                         "OpenCL") != std::string::npos)
+                    //     continue;
```
I think you didn't mean to push that.

hjabird replied:
I did not - thank you for spotting this. Reverted in [d0d10fe](https://github.com/oneapi-src/oneMKL/pull/411/commits/d0d10febbaf898e36bdd05f8579ef1a5338e0020).

Rbiessy commented on tests/unit_tests/dft/source/workspace_external_tests.cpp:
```diff
@@ -0,0 +1,402 @@
+        commit_descriptor(desc_usm, sycl_queue);
+    }
+    catch (oneapi::mkl::unimplemented&) {
+        // Admit defeat, and just test that the API does as expected.
```
Is there a reason this comment is different from the ones above in a similar situation?
```suggestion
        // DFT size may not be supported. Use a DFT size that probably will be, even if it
        // won't actually use an external workspace internally.
```

hjabird replied:
Applied in [985ea94](https://github.com/oneapi-src/oneMKL/pull/411/commits/985ea9406ed8cbb204e3bdff3b587df010374f1f).

hjabird committed:
Revert changes to test/unit_tests/main_test.cpp

hjabird committed:
Correct overrided to overridden

hjabird committed:
Apply Romain's suggestion to improve comment
